src/proper_target.erl

```diff
@@ -170,25 +173,55 @@ targeted(RawType) ->
 -spec init_target(proper_types:type()) -> ok.
 init_target(RawType) ->
   Type = proper_types:cook_outer(RawType),
+  case proper_types:find_prop(is_user_nf_stateful, Type) of
+    {ok, true} -> init_stateful();
+    {ok, false} -> ok;
+    error -> ok
+  end,
   TargetserverPid = get('$targetserver_pid'),
   safe_call(TargetserverPid, {init_target, Type}).
 
+%% Initialize stateful configuration.
+
+%% @private
+-spec init_stateful() -> ok.
+init_stateful() ->
+  TargetserverPid = get('$targetserver_pid'),
+  safe_call(TargetserverPid, init_stateful).
+
+%% Update the gen_server's process dictionary with some of
+%% PropEr's values in its process dictionary.
+
+%% @private
+-spec update_pdict([atom()]) -> ok.
+update_pdict(Keys) ->
+  update_pdict(Keys, []).
+
+%% @private
+update_pdict([], KVs) ->
+  TargetserverPid = get('$targetserver_pid'),
+  gen_server:call(TargetserverPid, {pdict, KVs});
+update_pdict([Key | Keys], KVs) ->
+  update_pdict(Keys, [{Key, get(Key)} | KVs]).
+
 %% This produces the next gen instance from the next
 %% generator provided by the strategy. It will also
 %% update the state and data of the strategy.
 
 %% @private
 -spec targeted_gen() -> any().
 targeted_gen() ->
+  update_pdict(['$left', '$size']),
   TargetserverPid = get('$targetserver_pid'),
   gen_server:call(TargetserverPid, gen).
 
 %% @doc Get the shrinker for a Type.
 
 -spec get_shrinker(proper_types:type()) -> proper_types:type().
 get_shrinker(Type) ->
-  TargetserverPid = get('$targetserver_pid'),
```

Comment:
This is not part of the changes, but I believe this needs to be refactored. Uses of get and put like this are wrong. This should be part of the propers run-state

Reply:
You are correct. However, at the moment the same thing is followed in `proper_typeserver`'s codebase. In order to update this, there would need to be a refactoring to `proper` module in order to work as a `gen_server` probably.